includes/Core/Util/Entity_Factory.php

```diff
@@ -560,4 +568,108 @@ private static function convert_to_amp_entity( $new_url, $entity ) {
 
 		return $new_entity;
 	}
+
+	/**
+	 * Gets the page number for a query, via the specified query var. Defaults to 1.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param WP_Query $query A WordPress query object.
+	 * @param string   $query_var Optional. Query var to look for, expects 'paged' or 'page'. Default 'paged'.
+	 * @return int The page number.
+	 */
+	private static function get_query_pagenum( $query, $query_var = 'paged' ) {
+		return $query->get( $query_var ) ? (int) $query->get( $query_var ) : 1;
+	}
+
+	/**
+	 * Paginates an entity URL.
+	 *
+	 * Logic extracted from `paginate_links` in WordPress core.
+	 * https://github.com/WordPress/WordPress/blob/7f5d7f1b56087c3eb718da4bd81deb06e077bbbb/wp-includes/general-template.php#L4203
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $url The URL to paginate.
+	 * @param int    $pagenum The page number to add to the URL.
+	 * @return string The paginated URL.
+	 */
+	private static function paginate_entity_url( $url, $pagenum ) {
```

Comment:
How about using something like this:

```php
$original_request_uri = $_SERVER['REQUEST_URI'];
$_SERVER['REQUEST_URI'] = str_replace( home_url(), '', $url );

$paginated_url = get_pagenum_link( $pagenum );

$_SERVER['REQUEST_URI'] = $original_request_uri;

return $paginated_url;
```

It is a bit hacky but does what we need without copy/pasting functionality from WordPress core.

Reply:
Thanks @eugene-manuilov, I would much prefer to reuse existing code too. Unfortunately, I did initially try reusing the logic from `get_pagenum_link`, but discovered in testing that it resulting in mismatched URLs in some cases when using the Plain Permalinks setting.

For example, with a paginated category, having navigated to page 2 via the pagination links, the URL is `http://sitekit.10uplabs.com/?paged=2&cat=3`.

![image](https://user-images.githubusercontent.com/18395600/152199797-3ec200ce-dcd7-42c4-9fd4-ef61e0caeccd.png)

But then, navigating via the Admin Bar to the Entity Dashboard:

![image](https://user-images.githubusercontent.com/18395600/152200350-854bea15-4d99-4104-9bc5-3ae4807511ba.png)

The URL for the page doesn't match, the order of the query params is reversed: https://sitekit.10uplabs.com/?cat=3&paged=2

![image](https://user-images.githubusercontent.com/18395600/152200927-7931d9b0-ac62-4203-89e7-8a56e87de0d4.png)

Using the logic from `paginate_links` results in the correct URL being shown in SiteKit. I realise there's a possibility that sometimes `get_pagenum_link` may be used for navigation instead but given the prominence of the paginated navigation I would think that's going to be the most common way for a user to get to a paginated page, although I could of course be mistaken.